front_end/webserver.py

```diff
@@ -145,20 +154,234 @@ def get(self):
+
+            new_admin = self.get_body_argument("new_admin")
+
+            if content.check_user_exists(new_admin):
+                if content.get_role(new_admin) == "administrator":
+                    result = f"{new_admin} is already an administrator."
+                else:
+                    content.add_admin_permissions(new_admin)
+                    result = f"Success! {new_admin} is an administrator."
+            else:
+                result = f"Error: The user '{new_admin}' does not exist."
+
+            self.render("profile_admin.html", page="admin", tab="manage", admins=content.get_users_from_role(0, "administrator"), result=result, user_info=content.get_user_info(user_id), user_logged_in=user_logged_in_var.get(), role=self.get_current_role())
+        except Exception as inst:
+            render_error(self, traceback.format_exc()) 
+
+class ProfileSelectCourseHandler(BaseUserHandler):
+    def get(self, user_id):
+        try:
+            role = self.get_current_role()
+            if role == "administrator" or role == "instructor":
+                courses = content.get_course_from_role(user_id)
+                if len(courses) > 1:
+                    self.render("profile_select_course.html", courses=courses)
+                else:
+                    self.render("profile_instructor.html", page="instructor", tab=None, course=courses[0][1], assignments=content.get_assignments(courses[0][1]['id']), assistants=content.get_users_from_role(courses[0][0], "assistant"), result=None, user_info=content.get_user_info(user_id), user_logged_in=user_logged_in_var.get(), role=self.get_current_role())
+        except Exception as inst:
+            render_error(self, traceback.format_exc()) 
+
+class ProfileInstructorHandler(BaseUserHandler):
+    def get (self, course_id, user_id):
+        try:
+            role = self.get_current_role()
+            if role == "administrator" or role == "instructor":
+                self.render("profile_instructor.html", page="instructor", tab=None, course=content.get_course_basics(course_id), assignments=content.get_assignments(course_id), assistants=content.get_users_from_role(course_id, "assistant"), result=None, user_info=content.get_user_info(user_id), user_logged_in=user_logged_in_var.get(), role=self.get_current_role())
+            else:
+                self.render("permissions.html", user_info=content.get_user_info(self.get_current_user()), user_logged_in=user_logged_in_var.get())                
+        except Exception as inst:
+            render_error(self, traceback.format_exc()) 
+
+    def post(self, course_id, user_id):
+        try:
+            role = self.get_current_role()
+            if role == "administrator" or role == "instructor":
+                new_assistant = self.get_body_argument("new_assistant")
+                result = ""
+
+                if content.check_user_exists(new_assistant):
+                    if content.get_role(new_assistant) == "assistant":
```

Comment:
Checking the role for an assistant (or an instructor) should be course specific. Please add an argument to this function that checks against the current course ID. Then update the code throughout webserver.py to deal with this properly. It'll have to deal with the fact that there is no course associated with an administrator role. I believe this is an old bug, unless I am forgetting something.

Reply:
I need some help with this one.